code/src/main/kotlin/com/expediagroup/sdk/core/interceptor/Interceptor.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2024 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.sdk.core.interceptor
+
+import com.expediagroup.sdk.core.http.Request
+import com.expediagroup.sdk.core.http.Response
+import java.io.IOException
+
+/**
+ * Represents an interceptor for modifying or augmenting HTTP requests and responses within the SDK.
+ *
+ * An `Interceptor` allows pre-processing of requests and post-processing of responses as they pass
+ * through the HTTP execution pipeline. This can be used for various purposes such as logging, authentication,
+ * retry logic, or adding custom headers.
+ */
+interface Interceptor {
```

Comment:
What do you think about renaming this to RequestInterceptor? Or did you have plans on making it more generic?

I was confused about LoggingInterceptor for a bit, thinking that it was for intercepting the logging, not the intercepting the request and logging that.

Reply:
Maybe, it's intercepting the `execute()` method actually. what could be a better name that indicates that this interceptor intercepts both the request and the response? `ExecutionInterceptor`?

Reply:
Execution of what though? `execute()` makes the request right?

Reply:
Do we need separate request/response interceptors?

Reply:
@jordan-n-schmidt Yes, but I would translate `RequestInterceptor` as something that runs before the request execution. In our, case, we are intercepting the response also.

@anssari1 Could be, but it will complicate things a bit, I'm with keeping it simple until we need more advanced usage. This is something internal to the SDK and not exposed to users, we can change it easily.

Reply:
I think request implies response, but it's fine since it's internal to sdk and we can easily change it.

Reply:
Some interceptions are unidirectional. e.g serialize request/deserialize response..

Reply:
@anssari1 This interface gives you the flexibility to do the interception logic before the request execution, after receiving the response, or both.

```kotlin
// Intercept before request execution
@Throws(IOException::class)
override fun intercept(chain: Interceptor.Chain): Response {
    val request = chain.request()

   // Process the request
    
    return chain.proceed(request)
}
```

```kotlin
// Intercept after request execution (on response)
@Throws(IOException::class)
override fun intercept(chain: Interceptor.Chain): Response {
    val response = chain.proceed(chain.request())

    // Process the response
    
    return response
}
```

```kotlin
// Intercept Both
@Throws(IOException::class)
override fun intercept(chain: Interceptor.Chain): Response {
    val request = chain.request()

    // Process the request

    val response = chain.proceed(request)
    
    // Process the response

    return response
}
```